exchange/replication.go

```diff
@@ -139,9 +139,12 @@ func NewReplication(h host.Host, idx *Index, dt datatransfer.Manager, rtv Routed
 
 	// TODO: clean this up
 	r.dt.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
-		if event.Code == datatransfer.Error && channelState.Recipient() == h.ID() {
-			// If transfers fail and we're the recipient we need to remove it from our index
-			r.idx.DropRef(channelState.BaseCID())
+		switch event.Code {
+		case datatransfer.Error:
+			if channelState.Recipient() == h.ID() {
+				// If transfers fail and we're the recipient we need to remove it from our index
+				r.idx.DropRef(channelState.BaseCID())
+			}
```

Comment:
```suggestion
			}
			case datatransfer.Complete:
			              // Probably can wrap the whole switch in this
			              if channelState.Recipient() == h.ID() {
			                              ref, err := idx.GetRef(channelState.BaseCID())
			                              store, err := idx.GetStore(ref.PayloadCID)
			                              keys, err := utils.IPLDMapKeys(context.TODO(), ref.PayloadCID, store.Loader)
			                              ref.Keys = keys // turn these into bytes
			                              idx.SetRef(ref)
			              }
```

Then in /internal/utils/state.go:
```

func MapKeys(ctx context.Context, root cid.Cid, loader ipld.Loader) ([]string, error) {
	// Turn the CID into an ipld Link interface, this will link to all the children
	lk := cidlink.Link{Cid: root}
	// Create an instance of map builder as we're looking to extract all the keys from an IPLD map
	nb := basicnode.Prototype.Map.NewBuilder()
	// Use a loader from the link to read all the children blocks from a given store
	err := lk.Load(ctx, ipld.LinkContext{}, nb, loader)
	if err != nil {
		return nil, err
	}
	// load the IPLD tree
	nd := nb.Build()
	// Gather the keys in an array
	entries := make([]string, nd.Length())
	it := nd.MapIterator()
	i := 0
	// Iterate over all the map entries
	for !it.Done() {
		k, _, err := it.Next()
		// all succeed or fail
		if err != nil {
			return nil, err
		}
		// The key IPLD node needs to be decoded as a string
		key, err := k.AsString()
		if err != nil {
			return nil, err
		}
		entries[i] = key
		i++
	}
	return entries, nil
}
```

Reply:
You can reuse that method in `idx.GetEntries()` checkout the tx.go file